File: src/main/java/rag/retrieval/DocumentStore.java
package rag.retrieval;

import rag.utils.JsonUtils;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class DocumentStore {

    private final List<Document> documents;

    private DocumentStore(List<Document> documents) {
        this.documents = documents;
    }

    public static DocumentStore load(String path) throws Exception {
        String raw = Files.readString(Path.of(path));
        List<Object> nodes = JsonUtils.expectArray(JsonUtils.parse(raw), "docs.json must be an array");
        List<Document> docs = new ArrayList<>();
        for (Object node : nodes) {
            Map<String, Object> obj = JsonUtils.expectObject(node, "Document entry must be an object");
            String id = obj.get("id").toString();
            String source = obj.getOrDefault("source", "Unknown").toString();
            String title = obj.getOrDefault("title", id).toString();
            String text = obj.getOrDefault("text", "").toString();
            docs.add(new Document(id, source, title, text));
        }
        return new DocumentStore(docs);
    }

    public List<Document> getDocuments() {
        return documents;
    }
}

File: src/main/python/rag/retrieval/document_store.py
from typing import List, Dict, Any
from pathlib import Path
from rag.retrieval.document import Document
from rag.utils.json_utils import JsonUtils


class DocumentStore:
    def __init__(self, documents: List[Document]) -> None:
        self.documents: List[Document] = documents

    @staticmethod
    def load(path: str) -> 'DocumentStore':
        raw: str = Path(path).read_text(encoding='utf-8')
        nodes: List[Any] = JsonUtils.expect_array(
            JsonUtils.parse(raw),
            "docs.json must be an array"
        )
        docs: List[Document] = []
        for node in nodes:
            obj: Dict[str, Any] = JsonUtils.expect_object(
                node,
                "Document entry must be an object"
            )
            doc_id: str = str(obj.get("id", ""))
            source: str = str(obj.get("source", "Unknown"))
            title: str = str(obj.get("title", doc_id))
            text: str = str(obj.get("text", ""))
            docs.append(Document(doc_id, source, title, text))
        return DocumentStore(docs)

    def get_documents(self) -> List[Document]:
        return self.documents






